src/server.rs

```diff
@@ -0,0 +1,221 @@
+            "DE" => Ok(DeleteEvidence),
+            "EE" => Ok(EditEvidence),
+            "ZZ" => Ok(CallModButton),
+            "opKICK" => Ok(KickWithGuard),
+            "opBAN" => Ok(BanWithGuard),
+            _ => anyhow::bail!("Invalid command!"),
+        }
+    }
+}
+
+enum DecodeState {
+    Command,
+    Arguments,
+}
+
+use serde::export::Formatter;
+use std::{
+    io::{BufRead, Cursor, Seek, SeekFrom},
+    str::FromStr,
+};
+use tokio::{
+    io::{AsyncBufReadExt, Error, ErrorKind},
+    stream::StreamExt,
+};
+use tokio_util::codec::{Decoder, Encoder, FramedRead};
+
+#[derive(Debug)]
+pub struct AOMessage {
+    pub command: Command,
+    pub args: Vec<String>,
+}
+
+pub struct AOMessageCodec {
+    state: DecodeState,
+}
+
+impl AOMessageCodec {
+    pub fn new() -> Self {
+        Self { state: DecodeState::Command }
+    }
+}
+
+impl Decoder for AOMessageCodec {
+    type Item = AOMessage;
+    type Error = anyhow::Error;
+
+    fn decode(
+        &mut self,
+        mut src: &mut BytesMut,
+    ) -> Result<Option<Self::Item>, Self::Error> {
+        let mut reader = Cursor::new(src);
+        let mut cmd_buf = Vec::with_capacity(8);
+        let cmd_len = BufRead::read_until(&mut reader, b'#', &mut cmd_buf)?;
+
+        if cmd_len == 0 {
+            log::debug!("Invalid protocol?!");
+            anyhow::bail!("Invalid protocol!")
+        }
+
+        cmd_buf.truncate(cmd_len.saturating_sub(1));
+        let command: Command = unsafe {
+            // We are sure that the commands will *always* be valid utf8
+            String::from_utf8_unchecked(cmd_buf)
+        }
+        .parse()?;
+
+        reader.seek(SeekFrom::Start(cmd_len as u64))?;
+        let mut splitted = BufRead::split(reader, b'#')
+            .map(|l| String::from_utf8(l.unwrap()).unwrap());
+        let mut args: Vec<String> = splitted.collect();
+        let mut end_header_index = 0;
+
+        for (i, arg) in args.iter().enumerate() {
+            if arg.chars().nth(0).unwrap() == '%' {
+                end_header_index = i;
+                break;
+            } else {
+                continue;
+            }
+        }
+
+        args.remove(end_header_index);
+
+        Ok(Some(AOMessage { command, args: vec![] }))
+    }
+}
+
+fn clean_utf8(mut bytes: &[u8]) -> String {
+    let mut output = String::new();
+
+    loop {
+        match str::from_utf8(bytes) {
+            Ok(s) => {
+                output.push_str(s);
+                return output;
+            }
+            Err(e) => {
+                let (good, bad) = bytes.split_at(e.valid_up_to());
+
+                if !good.is_empty() {
+                    let s = unsafe { str::from_utf8_unchecked(good) };
+                    output.push_str(s);
+                }
+
+                if bad.is_empty() {
+                    return output;
+                }
+
+                output.push_str("");
+
+                bytes = &bad[1..];
+            }
+        }
+    }
+}
```

Comment:
This all is just  [`from_utf8_lossy`](https://doc.rust-lang.org/alloc/string/struct.String.html#method.from_utf8_lossy) + `.replace("�", "")`